src/OpenTelemetry/Metrics/HistogramBuckets.cs

```diff
@@ -36,9 +40,38 @@ public class HistogramBuckets
 
         internal int IsCriticalSectionOccupied = 0;
 
+        private readonly BucketLookupNode bucketLookupTreeRoot;
+
+        private readonly Func<double, int> findHistogramBucketIndex;
+
         internal HistogramBuckets(double[] explicitBounds)
         {
             this.ExplicitBounds = explicitBounds;
+            this.findHistogramBucketIndex = this.FindBucketIndexLinear;
+            if (explicitBounds != null && explicitBounds.Length >= DefaultBoundaryCountForBinarySearch)
+            {
+                this.bucketLookupTreeRoot = ConstructBalancedBST(explicitBounds, 0, explicitBounds.Length);
+                this.findHistogramBucketIndex = this.FindBucketIndexBinary;
+
+                static BucketLookupNode ConstructBalancedBST(double[] values, int min, int max)
```

Comment:
I'd like to reopen the discussion over binary search vs binary search tree, but the old thread is on outdated code and hidden: https://github.com/open-telemetry/opentelemetry-dotnet/pull/3252#discussion_r930535468